### DISCUSSION TITLE
Laplacian of a scalar field

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#23551] [Q&A General]
Hello everyone,

Is there anyway in MOOSE to get the Laplacian of a scalar field?

I need to compute d2c/dx2 but do know how to do it.

Thanks,
Ali

**Comment by GiudGiud** [comment]
Hello

By scalar field you mean a regular variable?
Do you mean to use this in a residual or just for postprocessing?

Guillaume

**Reply by alimostafavi24** [reply]
Yes, I meant a regular variable. I need it for post processing.

**Reply by GiudGiud** [reply]
You can look at second derivatives on variables (retrieve using getVar())
https://mooseframework.inl.gov/docs/doxygen/moose/classMooseVariableFE.html#a6084c139e379961bfcc3f5ed97d181f4
or use the Coupleable interface, same as for values, first derivatives etc
https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html#a905ce7bd997b1374a18b0c7a3d3389a5

**Reply by alimostafavi24** [reply]
Thank you so much. In one part of my code, I also need the gradient of the component to compute the residual. I am not getting the correct result since I am using VariableGradientComponent. In this way the variable is totally constant inside the element. Is there any way to fix this?

Thanks,
Ali

**Reply by GiudGiud** [reply]
use coupledGradient to retrieve a gradient within your kernel.

Do not use an auxiliary variable to store a gradient or a second derivative for anything else than postprocessing.

**Reply by alimostafavi24** [reply]
I have a problem in calculating the residual correctly. Here is my .h and .C file.


```
#pragma once

#include "Kernel.h"

class SurfaceTension : public Kernel
{
public:
  SurfaceTension(const InputParameters & parameters);

  static InputParameters validParams();

protected:
  virtual Real computeQpResidual() override;

  virtual Real computeQpJacobian() override;

private:
  unsigned int _v_var;
  const VariableGradient & _grad_c;
  const VariableValue & _w;
  Real _coef;
};
```

```

#include "SurfaceTension.h"

registerMooseObject("ExampleApp", SurfaceTension);

InputParameters
SurfaceTension::validParams()
{
  InputParameters params = Kernel::validParams();

  params.addRequiredCoupledVar(
      "w", "The gradient of this variable will be used.");
    params.addRequiredCoupledVar(
      "c", "The value of this variable will be used.");
      params.addParam<Real>(
      "coef", 1.0, "Coefficent ($\\sigma$) multiplier for the coupled force term.");


  return params;
}

SurfaceTension::SurfaceTension(const InputParameters & parameters)
  : Kernel(parameters),

    _v_var(coupled("v")),
    _grad_c(coupledGradient("c")),
    _w(coupledValue("w"))
{
}

Real
SurfaceTension::computeQpResidual()
{

  return _test[_i][_qp] * (_grad_c[_qp] * _w[_qp]) * _coef;

}

Real
SurfaceTension::computeQpJacobian()
{

  return _test[_i][_qp] * (_grad_c[_qp] * _w[_qp] * _grad_phi[_j][_qp]) * _coef;
}
```
The error that I get while compiling is as below.

```
/home/ali/projects/moose/modules/navier_stokes/src/kernels/SurfaceTension.C: In member function 'virtual libMesh::Real SurfaceTension::computeQpResidual()':
/home/ali/projects/moose/modules/navier_stokes/src/kernels/SurfaceTension.C:48:52: error: cannot convert 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'} to 'libMesh::Real' {aka 'double'} in return
   48 |   return _test[_i][_qp] * (_grad_c[_qp] * _w[_qp]) * _coef;
make: *** [/home/ali/projects/moose/framework/build.mk:145: /home/ali/projects/moose/modules/navier_stokes/build/unity_src/kernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```

I would highly appreciate it if you can help me to compute the residual for this source term which essentially is a vector.

Thanks,
Ali

**Reply by GiudGiud** [reply]
The source term is a vector?
Is this contributing to a vector equation (conservation of momentum?)

or a regular scalar equation.

Are your variables vector variables? What are the variables?

**Comment by alimostafavi24** [comment]
w and c are both scalar variables. Therefore w*grad(c) is a vector value and this is a source term for a vector equation.

Thanks,
Ali

**Reply by GiudGiud** [reply]
what is this equation if you dont mind writing it down

what are you planning to use as the main variable for this equation

**Reply by alimostafavi24** [reply]
<img width="404" alt="Screen Shot 2023-02-27 at 7 50 03 PM" src="https://user-images.githubusercontent.com/119982603/221741074-c9d7da62-8fff-4cb3-88e1-9bcb4cbb9748.png">
<img width="223" alt="Screen Shot 2023-02-27 at 7 50 09 PM" src="https://user-images.githubusercontent.com/119982603/221741100-8e2968b8-e3cf-4b34-8480-60d5bb0c5c6c.png">
<img width="83" alt="Screen Shot 2023-02-27 at 7 50 23 PM" src="https://user-images.githubusercontent.com/119982603/221741111-5c8d1b65-1cbc-497f-a565-2b0023da44ed.png">

w and c are both computed using phase field kernel.

**Reply by GiudGiud** [reply]
ok this top equation is indeed a vector equation, it s the conservation of momentum.

What is u in your simulation? Are you using Navier Stokes? and which discretization within Navier Stokes?
This is important to advise you on which type of kernel to use to contribute to this equation

**Reply by alimostafavi24** [reply]
I am using INSMassRZ, INSMomentumTimeDerivative and INSMomentumLaplaceFormRZ. In addition, my variables in momentum equation are velocity by its components (vel_x, vel_y) and pressure.

Thanks,
Ali

**Reply by GiudGiud** [reply]
in that context, you want to use an integration by parts instead of computing the Laplacian
